### ISSUE TITLE
How could I connect the Unity's Texture2D with Android Studio's ExoPlayer?

### ISSUE THREAD

**Issue by starias** [google/ExoPlayer#10915]
I'm learning about creating a plugin with [ExoPlayer](https://exoplayer.dev/) to render video-streams in Unity. I know this is not an issue directly related with ExoPlayer, but I'm searching for help. I have this same issue posted in [stackoverflow](https://stackoverflow.com/questions/75057101/how-to-create-android-view-textureview-from-unitys-texture2d).

I have the exoplayer working fine, everything is connected and I'm using also [JNI](https://en.wikipedia.org/wiki/Java_Native_Interface) to build the wrapper between Unity and Andriod native. The player is playing and I have sound, I can pause, play, seek, etc, but I don't have the rendering. 

My idea is creating a TextureView or a SurfaceView from Unity's Texture2D, but I don't know exactly how. As they explain in the documentation and also as I saw, ExoPlayer needs to be provided by one of those SurfaceView or TextureView. 

In AndroidStudio it's easy and it doesn't need too much because you can create a SurfaceView directly from the .xml off the app, but in Unity things are pretty different.

I'm trying to extrapolate the code from [unity-android-native-camera sample](https://github.com/robsondepaula/unity-android-native-camera) in order to create a Texture2D and use the pointer of that texture as they do. I'm not doing the exactly same things from the repo, but the concept is the same.

**Unity code**  
- **CreateTextureAndPassToPlugin** is used to create the Texture2D pointer, apply it to the RawImage or Material and the pass that pointer to the JNI to save the value.
-  **CallPluginAtEndOfFrames** is used to update the texture pointer from the Android part. 
- **InitExoPlayer** initializes ExoPlayer pressing a button in the scene.

Anyway, I don't think these 2 functions are doing anything, because in Android part I tried different things and the current version is trying to create a TextureView and set it into the ExoPlayer instance, because if you do that ExoPlayer-setTextureView directly in Android Studio (with no Unity) the rendering is updated automatically. But those functions are copy from the repo I mentioned before.

```C#
...
    void Start()
    {
        for (int i = 0; i < samples.Count; i++)
        {
            stringList.Add("");
        }

        try
        {
            samples[0].text = stringList[0] = getDebugLog();
            samples[1].text = stringList[1] = "TEXTURE: " + CreateTextureAndPassToPlugin();
        }
        catch (Exception e)
        {
            Console.WriteLine(e);
            throw;
        }

        StartCoroutine(nameof(CallPluginAtEndOfFrames));
    }
...

    private IEnumerator CallPluginAtEndOfFrames()
    {
        while (true)
        {
            // Wait until all frame rendering is done
            yield return new WaitForEndOfFrame();

            // Issue a plugin event with arbitrary integer identifier.
            // The plugin can distinguish between different
            // things it needs to do based on this ID.
            // For our simple plugin, it does not matter which ID we pass here.
            GL.IssuePluginEvent(updateTexture(), 1);

            // skip one frame
            yield return new WaitForEndOfFrame();
        }
    }
...

    public void InitExoPlayer()
    {
        if (!initExoButton.enabled)
            return;

        // ExoPlayer init
        createUnityPlayer(uri, playing, enableRendering);
        ...
    }
... 

    private bool CreateTextureAndPassToPlugin()
    {
        Texture2D tex = new Texture2D(960, 640, TextureFormat.RGBA32, false);
        // Set point filtering just so we can see the pixels clearly
        tex.filterMode = FilterMode.Point;
        // Call Apply() so it's actually uploaded to the GPU
        tex.Apply();

        //displayMaterial.mainTexture = tex;
        displayRawImage.texture = tex;

        return setTextureFromUnity(tex.GetNativeTexturePtr());
    }

...
```

**JNI code**
The pointer of the Texture2D is saved into **g_TexturePointer**, so then I can use it in different functions. At the beginning I was using it as the repo uses it, but now I'm using it to initialize ExoPlayer with it, because I use that pointer to create a **SurfaceTexture** and then a **TextureView**.

```C++
...
extern "C" void UNITY_INTERFACE_EXPORT UNITY_INTERFACE_API
createUnityPlayer(char *uri, bool isAutoPlay, bool isRendering) {
    JNIEnv *jni_env_local = getJNIEnv();
    if (!jni_env_local || !EXO_createUnityPlayer) {
        releaseJNIEnv();
        return;
    }

    jstring sUri = jni_env_local->NewStringUTF(uri);
    if (g_TexturePointer) {
        GLuint gltex = (GLuint) (size_t) (g_TexturePointer);
        jni_env_local->CallVoidMethod(exoBridgeObject, EXO_createUnityPlayer,
                                      sUri, isAutoPlay, isRendering, (int) gltex);
    } else {
        jni_env_local->CallVoidMethod(exoBridgeObject, EXO_createUnityPlayer,
                                      sUri, isAutoPlay, isRendering, -1);
    }
}
...
static void UNITY_INTERFACE_API OnRenderEvent(int eventID) {
    JNIEnv *jni_env_local = getJNIEnv();
    if (!jni_env_local || !EXO_updateTexture || !g_TexturePointer) {
        releaseJNIEnv();
        return;
    }
    GLuint gltex = (GLuint) (size_t) (g_TexturePointer);
    jni_env_local->CallVoidMethod(exoBridgeObject, EXO_updateTexture, (int) gltex);
    releaseJNIEnv();
}

extern "C" UnityRenderingEvent UNITY_INTERFACE_EXPORT UNITY_INTERFACE_API
updateTexture() {
    return OnRenderEvent;
}
...
```

**ExoActivity.java**
I'm creating a new Activity which extends from UnityPlayerActivity. The **updateTexture function** is used to reproduce the steps from the repo, but as I said before I think this code is not relevant right now. Inside the **createUnityPlayer function** it is called **prepareExoRender function** in order to set-up the ExoPlayer's TextureView using the pointer of the Texture2D from Unity.

```Java
...
    public void createUnityPlayer(String uri, boolean isAutoPlay,
                                  boolean isRendering, int texturePointer) {
        ...
        prepareExoRender(texturePointer);
        ...
    }
...
    TextureView _exoView;
    SurfaceTexture _exoTex;
    /**
     *
     */
    private void prepareExoRender(int texturePointer) {
        ...
        try {
            _exoView = new TextureView(this);
            _exoTex = new SurfaceTexture(texturePointer);
            _exoView.setSurfaceTexture(_exoTex);

            _exoPlayer.prepareUnityView(_exoView);
            _debugLog = "SUCCESS-TEXTURE_VIEW";
        } catch (Exception e) {
            _debugLog = "- " + e.getMessage();
        }
    }
...
    private void updateTexture(int texturePointer) {
        if (!_isRendering)
            return;

        int[] imageBuffer = new int[0];

        if (_conversionScript != null) {
            imageBuffer = _conversionScript.getOutputBuffer();
        }

        if (imageBuffer.length > 1) {
            GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, texturePointer);

            GLES20.glTexSubImage2D(GLES20.GL_TEXTURE_2D, 0, 0, 0, _previewSize.getWidth(),
                    _previewSize.getHeight(), GLES20.GL_RGBA, GLES20.GL_UNSIGNED_BYTE,
                    IntBuffer.wrap(imageBuffer));
        }
        _exoTex.updateTexImage();
    }
...
```

**ExoClass.java**
Class which contains all the information related with ExoPlayer, basically the instance and related.

```Java
...
    public void prepareUnityView(TextureView view) {
        if (view == null || _player == null) {
            _result = "NON-SURFACE-CREATED";
            return;
        }
        try {
            _result = "SET_VIDEO_SURFACE";
            _player.setVideoTextureView(view);
            _result = "SUCCESS - VIEW";
        } catch (Exception e) {
            _result += "- " + e.getMessage();
        }
    }
...
```
As I tested in an raw-AndroidStudio project, if I create a TextureView and then I do the following:
```
TextureView playerView = findViewById(R.id.textureView);
        videoPlayer.setPlayerDefaultView(playerView);
        videoPlayer.prepareVideoPlayer("https://storage.googleapis.com/wvmedia/clear/h264/tears/tears.mpd");
        videoPlayer.startPlayback();
```
the player will be rendered automatically every frame.

How could I create a TextureView correctly?
Is there other ways to do this?

**Comment by andrewlewis** [comment]
It should be possible to [instantiate a `Surface` from the `SurfaceTexture`](https://developer.android.com/reference/android/view/Surface#Surface(android.graphics.SurfaceTexture)) and then pass the `Surface` to ExoPlayer (you don't need to pass a view, although for use cases where you do pass a view then the player takes care of handling lifecycle events on the surface). I don't think a `TextureView` makes sense for this use case.

There are a couple of older issues on this tracker about using ExoPlayer in a Unity plugin, and they link to this [blog post](https://medium.com/cinemur-engineering/exoplayer-for-building-powerful-vr-players-for-cardboard-and-gearvr-73ec7e83dd5c#.52ou4a6su). Not sure if it's still applicable though (this isn't really something we can provide support on).

**Comment by starias** [comment]
Ok, fine.  I checked the link you provided me and also I've been watching different post about creating a SurfaceTexture in AndroidStudio with OpenGL and now I have to try different things until it works. 

The point is I was taking this in the wrong way. I was trying to do the same as I did into the native AndroidStudio's project, where I used the SurfaceView or TextureView with ExoPlayer and I didn't have to do anything else. 

Within Unity, it works different, because even though ExoPlayer is working from inside, the render is something that has to be updated in Unity's side too. The way to do this (as I have understood) is making the SurfaceTexture in AndroidStudio, then the SurfaceTexture has to be attached with the Unity's Texture2D you want to use and then it is necessary to call the function "[updateTexImage()](https://developer.android.com/reference/android/graphics/SurfaceTexture#updateTexImage())" to  render the ExoPlayer's content into the Texture2D.  Also it is needed to create a Surface with the SurfaceTexture I mentioned and attach that Surface with ExoPlayer (as you told me before).

In addition, this should be called within coroutines of Unity, because it's more efficient and less dangerous in runtime. So all I need right now is learn how to create and how to use the OpenGL stuff in order to do all of this things.

Thank you very much! You gave me a new perspective about how to approach this problem.
